### DISCUSSION TITLE
Implementing DivideMix

### DISCUSSION THREAD

**Discussion by dreaquil** [Lightning-AI/pytorch-lightning#10561] [code help: CV]
Hi! Recently started using pytorch lightning and really enjoying it so far! However, I'm still getting used to the way of doing things. I'm currently working on implementing  a [DivideMix](https://arxiv.org/pdf/2002.07394.pdf) model which is based on MixMatch (Semi supervised learning). This has a rather complicated training loop as it contains two models and uses the output of one model to sample the training data for the other. I'm struggling to figure out how best to fit this into the pytorch lightning framework or if this may be beyond the framework's structure. Does anyone have any idea what the best way to structure this would be?

Training starts with:
- A warm up period for both models (first 10 epochs or so).

From my understanding, the following steps need to take place on each epoch:

- Loss calculation for each sample in the full training set by both models.
- Fitting a two component GMM to the loss distribution for each model.
- Using the GMM produced by model 2 to split training data into clean and noisy data for training model 1
- Label refinement/ Label Co-guessing for model 1 inputs (Still a bit hazy)
- Mixing inputs
- Optimization for model 1
- Using the GMM produced by model 1 to split training data into clean and noisy data for training model 2
- Label refinement / Label Co-guessing for model 2 inputs (Still a bit hazy)
- Mixing inputs
- Optimization for model 2
